[LT] Add IR resuing support for manually-implemented ops

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #77616

Summary: Add CanBeReused methods for manually-implemented ops and replace MakeNode with
ReuseOrMakeNode.

desertfire committed:
[LT] Add IR resuing support for manually-implemented ops

Summary: Add CanBeReused methods for manually-implemented ops and replace MakeNode with
ReuseOrMakeNode.

[ghstack-poisoned]

desertfire commented:
@JackCaoG , this PR only touches ops for ts_backend, but can answer your earlier question about how to handle manually-implemented ops.

JackCaoG reviewed (approved):
Thanks!

desertfire committed:
Update on "[LT] Add IR resuing support for manually-implemented ops"


Summary: Add CanBeReused methods for manually-implemented ops and replace MakeNode with
ReuseOrMakeNode.

[ghstack-poisoned]

wconstab commented on torch/csrc/lazy/ts_backend/ops/scalar.h:
```diff
@@ -19,6 +19,16 @@ class TORCH_API Scalar : public TsNode {
   Scalar(const at::Scalar& value, c10::ScalarType type);
 
+  bool CanBeReused(const at::Scalar& value, Shape shape) const {
+    TORCH_INTERNAL_ASSERT(false, "Reusing Scalar nodes is unsupported")
```
hmm, so, i guess you actually prevent this by not calling ReuseOrMakeNode on scalar, but this is just a backup in case someone tried to change that?

wconstab commented on torch/csrc/lazy/ts_backend/ir_builder.h:
```diff
@@ -36,28 +36,30 @@ namespace lazy {
-  NodePtr MakeView(const Value& input0, const std::vector<int64_t>& output_size) const override { return MakeNode<View>(input0, output_size); }
-  NodePtr MakeCast(const Value& input0, const at::ScalarType& dtype, const c10::optional<at::ScalarType>& stype = c10::nullopt) const override { return MakeNode<Cast>(input0, dtype, stype); }
-  NodePtr MakeTensorList(const OpList& inputs) const override { return MakeNode<TensorList>(inputs); }
+  NodePtr MakeScalar(const at::Scalar& value, const at::ScalarType& type) const override { return ReuseOrMakeNode<Scalar>(value, type); }
```
hmm, no you actually do change it here. so why do you assert in the CanBeReused for scalar?

wconstab reviewed (approved):
lgtm pending the one question i left

desertfire committed:
Update on "[LT] Add IR resuing support for manually-implemented ops"


Summary: Add CanBeReused methods for manually-implemented ops and replace MakeNode with
ReuseOrMakeNode.

[ghstack-poisoned]

desertfire committed:
Update on "[LT] Add IR resuing support for manually-implemented ops"


Summary: Add CanBeReused methods for manually-implemented ops and replace MakeNode with
ReuseOrMakeNode.

[ghstack-poisoned]

desertfire committed:
Update on "[LT] Add IR resuing support for manually-implemented ops"


Summary: Add CanBeReused methods for manually-implemented ops and replace MakeNode with
ReuseOrMakeNode.

[ghstack-poisoned]

desertfire commented:
@pytorchbot merge this
